### ISSUE TITLE
Refactor the generic intersection observer part of amp-analytics/visibility-impl.js into a separate service.

### ISSUE THREAD

**Issue by aghassemi** [ampproject/amphtml#4780]
Currently `VideoManager` needs to be notified when x% of an element becomes visible in the viewport. 
`amp-analytics/visibility-impl.js` is already doing this by listening to `onScroll` and `onChanged` but the code is mixed with other analytics code.

`intersection-observer.js` (unfortunately named that collides with native `IntersectionObserver`) also listens to `onScroll` and `onChanged` and does a similar thing.

We should:
- [ ]  Have a service that, as much as possible, mimics the **native IntersectionObserver** (https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)
- [ ] Actually use native IntersectionObserver when supported and polyfill using the official IntersectionObserver polyfill (https://github.com/WICG/IntersectionObserver/tree/gh-pages/polyfill) or if it does not fit our needs, polyfill with viewer `onScroll` and `onChanged` as we do now.
- [ ] Use the new `IntersectionObserver` service in `VideoManager`
- [ ] Use the new `IntersectionObserver` service in `amp-analytics/visibility-impl.js`
- [ ] Use the new `IntersectionObserver` service in `iframe` and `ads` and remove our `intersection-observer.js` that is tied specifically to `iframe` and `ads`.

**Comment by jridgewell** [comment]
> Rename our intersection-observer.js to iframe-intersection-observer.js

Let's just fix up iframes and ads to use a real intersection observer. Re: https://github.com/ampproject/amphtml/issues/4048#issuecomment-243551760

**Comment by aghassemi** [comment]
@jridgewell hehe, today is all about IntersectionObserver :hotsprings: 
Updated the description to match your comment.

**Comment by ericlindley-g** [comment]
@aghassemi  Putting into "pending triage", but please update the milestone when you have a sense of when work will start—thanks!

**Comment by aghassemi** [comment]
/to @muxin for after #6398 since you would know everything about the new intersection observer by then :)

**Comment by aghassemi** [comment]
well this is actually done by @zhouyx, the part I want to do is using it in video-manager. Will create a separate issue.
